hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/util/PendingAirdropUpdater.java

```diff
@@ -0,0 +1,125 @@
+import com.hedera.node.app.service.token.impl.WritableAirdropStore;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Class that computes the expected results of a pending airdrops removal and then commit the needed changes.
+ * This way it will save several state updates of pending airdrop's linked list pointers {@code previousAirdrop()}
+ * and {@code nextAirdrop()}, also sender's account updates of {@code headPendingAirdropId()} and
+ * {@code numberPendingAirdrops()}
+ */
+public class PendingAirdropUpdater {
+    @NonNull
+    private final WritableAirdropStore pendingAirdropStore;
+
+    @NonNull
+    private final WritableAccountStore accountStore;
+
+    public PendingAirdropUpdater(
+            @NonNull WritableAirdropStore pendingAirdropStore, @NonNull WritableAccountStore accountStore) {
+        this.pendingAirdropStore = pendingAirdropStore;
+        this.accountStore = accountStore;
+    }
+
+    public void removePendingAirdrops(@NonNull final List<PendingAirdropId> airdropsToRemove) {
+        final Map<AccountID, Account> updatedSenders = new HashMap<>();
+        final Map<PendingAirdropId, AccountPendingAirdrop> updatedAirdrops = new HashMap<>();
+        // calculate state changes
+        for (PendingAirdropId id : airdropsToRemove) {
+            removePendingAirdropAndUpdateStores(id, updatedSenders, updatedAirdrops);
+        }
+
+        // commit updates
+        updatedSenders.forEach((accountID, account) -> accountStore.put(account));
+        updatedAirdrops.forEach(pendingAirdropStore::put);
+        airdropsToRemove.forEach(pendingAirdropStore::remove);
+    }
+
+    private void removePendingAirdropAndUpdateStores(
+            final PendingAirdropId airdropId,
+            final Map<AccountID, Account> updatedSenders,
+            final Map<PendingAirdropId, AccountPendingAirdrop> updatedAirdrops) {
+
+        final var senderId = airdropId.senderIdOrThrow();
+        var senderAccount = updatedSenders.containsKey(senderId)
+                ? updatedSenders.get(senderId)
+                : requireNonNull(accountStore.getAccountById(senderId));
+
+        final var airdrop = updatedAirdrops.containsKey(airdropId)
+                ? updatedAirdrops.get(airdropId)
+                : pendingAirdropStore.getForModify(airdropId);
+        validateTrue(airdrop != null, INVALID_TRANSACTION_BODY);
+
+        // updated sender's head of pending airdrops
+        if (airdropId.equals(senderAccount.headPendingAirdropId())) {
+            final var updatedSender = senderAccount
+                    .copyBuilder()
+                    .headPendingAirdropId(airdrop.nextAirdrop())
+                    .build();
+            updatedSenders.put(updatedSender.accountId(), updatedSender);
+        }
+
+        // update pending airdrops links
+        final var prevAirdropId = airdrop.previousAirdrop();
```

Comment:
Shouldn't all the below be in else condition ?

Reply:
I did a refactoring on this method.